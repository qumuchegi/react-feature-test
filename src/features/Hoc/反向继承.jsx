import React from 'react'

export class A extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      a: 0
    }
  }
  handleClick = () => {
    this.setState(({a}) => ({
      a: a+1
    }))
  }
  render () {
    return <div>
      <div>props.b: {this.props.b}</div>
      <div id="state-a">{this.state.a}</div>
      <div onClick={this.handleClick} id='button'>点击</div>
    </div>
  }
}

const Hoc = (Warped) => class extends Warped {
  handleClick = () => {
    console.log(this.state)
    super.setState(({a}) => ({
      a: a+2
    }))
  }

  render () {
    return super.render();
  }
}

export default A//Hoc(A)