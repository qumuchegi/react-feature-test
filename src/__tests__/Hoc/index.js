import React from 'react'
import TestRenderer from 'react-test-renderer'
import ReactTestUtils from 'react-dom/test-utils'
import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import { A } from '../../features/Hoc/反向继承.jsx'

Enzyme.configure({ adapter: new Adapter() })

// describe('A component', () => {
//   let warper 
//   let prevStateA

//   beforeEach(() => {
//     warper = TestRenderer.create(<A />)
//     prevStateA = warper.root.instance.state.a
//   })
//   test('default props', () => {
//     const tree = warper.toJSON()
//     expect(tree).toMatchSnapshot()
//   })
//   test('should increase 1 after click button', () => {
//     const instance = warper.getInstance()
//     const button = warper.root.findByProps({ id: 'button' })
//     const showStatA = warper.root.findByProps({id: 'state-a'})
//     instance.handleClick()
//     expect(button.children[0]).toBe('点击')
//     expect(showStatA.children[0]).toBe('1')
//     expect(instance.state.a).toBe(prevStateA + 1)
//   })
// })

// use anzyme
describe('A component', () => {
  let warper
  let prevStateA
   
  beforeEach(() => {
    warper = shallow(<A/>)
    prevStateA = warper.state().a
  })

  test('default props', () => {
    //const tree = warper.toJSON()
    expect(warper).toMatchSnapshot()
  })
  test('should increase 1 after click button', () => {
    const button = warper.find('#button')
    button.simulate('click')
    warper.update() // simulate 之后要手动更新才能拿到子组件的新文本
    const showStatA = warper.find('#state-a')
    expect(button.text()).toBe('点击')
    expect(showStatA.text()).toBe('1')
    expect(warper.state().a).toEqual(prevStateA + 1)
  })

})