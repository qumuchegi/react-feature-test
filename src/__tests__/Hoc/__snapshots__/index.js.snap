// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`A component default props 1`] = `ShallowWrapper {}`;
